streaming/src/main/scala/org/apache/spark/streaming/util/BatchedWriteAheadLog.scala

```diff
@@ -0,0 +1,212 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.streaming.util
+
+import java.nio.ByteBuffer
+import java.util.concurrent.{LinkedBlockingQueue, TimeoutException}
+import java.util.{Iterator => JIterator}
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable.ArrayBuffer
+import scala.concurrent.{Await, ExecutionContext, Promise}
+import scala.concurrent.duration._
+import scala.util.control.NonFatal
+
+import org.apache.spark.Logging
+import org.apache.spark.util.{Utils, ThreadUtils}
+
+/**
+ * A wrapper for a WriteAheadLog that batches records before writing data. All other methods will
+ * be passed on to the wrapped class.
+ *
+ * Parent exposed for testing.
+ */
+private[streaming] class BatchedWriteAheadLog(private[util] val parent: WriteAheadLog)
+  extends WriteAheadLog with Logging {
+
+  import BatchedWriteAheadLog._
+
+  // exposed for tests
+  protected val walWriteQueue = new LinkedBlockingQueue[RecordBuffer]()
+
+  private val WAL_WRITE_STATUS_TIMEOUT = 5000 // 5 seconds
+
+  // Whether the writer thread is active
+  @volatile private var active: Boolean = true
+  protected val buffer = new ArrayBuffer[RecordBuffer]()
+
+  private val batchedWriterThread = startBatchedWriterThread()
+
+  /**
+   * Write a byte buffer to the log file. This method adds the byteBuffer to a queue and blocks
+   * until the record is properly written by the parent.
+   */
+  override def write(byteBuffer: ByteBuffer, time: Long): WriteAheadLogRecordHandle = {
+    val promise = Promise[WriteAheadLogRecordHandle]()
+    walWriteQueue.offer(RecordBuffer(byteBuffer, time, promise))
+    try {
+      Await.result(promise.future.recover { case _ => null }(ThreadUtils.sameThread),
+        WAL_WRITE_STATUS_TIMEOUT.milliseconds)
+    } catch {
+      case e: TimeoutException =>
+        logWarning(s"Write to Write Ahead Log promise timed out after " +
+          s"$WAL_WRITE_STATUS_TIMEOUT millis for record.")
+        null
+    }
+  }
+
+  /**
+   * Read a segment from an existing Write Ahead Log. The data may be aggregated, and the user
+   * should de-aggregate using [[BatchedWriteAheadLog.deaggregate]]
+   *
+   * This method is handled by the parent WriteAheadLog.
+   */
+  override def read(segment: WriteAheadLogRecordHandle): ByteBuffer = {
+    parent.read(segment)
+  }
+
+  /**
+   * Read all the existing logs from the log directory.
+   *
+   * This method is handled by the parent WriteAheadLog.
+   */
+  override def readAll(): JIterator[ByteBuffer] = {
+    parent.readAll().asScala.flatMap(deaggregate).asJava
```

Comment:
`flatMap` does not, but the `asScala` and `asJava` calls are allocating new objects. They maybe lazily collecting the objects being iterated over, but they are still creating two new objects no? In a loop, the only new object being created is the list itself. The others are basically references to already created objects returned from the `deaggregate()` call.

It might be more code, but I never found loops hard to read - sometimes I find loops clearer. 

I don't feel that strongly about it - I just feel that a couple of scala to java (and vice versa) conversions just for a `flatMap()` call is not really worth it. Being Young-gen GC they are not that bad, I guess.